### PR TITLE
Python 3 support for Urban Access

### PR DESCRIPTION
**Pull request by kuanb**
Updated: Should be ready for a review. Rebased with master as well, resolving merge conflicts that were outstanding. Let me know if you have any questions.

Needs tests, but that's a much larger undertaking, involving adding tests for the entire lib... Happy to help on that.

### CODE REVIEW

**Comment by pksohn** [comment]
Thanks. Will review this next week.

**Comment by bhelx** [comment]
This branch worked for me but required installing `future`.

```
pip3 install future
```

Perhaps you want to add that to the dependencies list in the readme and the setup.py.

File: urbanaccess/gtfs/network.py [hunk 1]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
+
+        time_diff_interp_stops = timedifference(stop_times_df=interp_stops)
+        gtfsfeeds_df.stop_times_int = time_diff_interp_stops
+
+        # check if we want to save this output to hdf5 file
         if save_processed_gtfs:
-            save_processed_gtfs_data(gtfsfeeds_dfs=gtfsfeeds_dfs,
-                                     dir=save_dir, filename=save_filename)
+            save_processed_gtfs_data(
+                                gtfsfeeds_df=gtfsfeeds_df,
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: I think a search and replace may have caused a mistake in your variable naming. I think everywhere you have `gtfsfeeds_df` it should be `gtfsfeeds_dfs`.

File: urbanaccess/gtfs/network.py [hunk 2]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
         gtfsfeeds_dfs.stop_times_int = _time_difference(
             stop_times_df=gtfsfeeds_dfs.stop_times_int)
 
+    # direction_id is conditional, though
+    if 'direction_id' not in gtfsfeeds_df.trips.columns:
+        columns.remove('direction_id')
+    
+    trips_subgdf = gtfsfeeds_df.trips[columns]
+    calendar_selected_trips_df = tripschedualselector(
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: `tripschedualselector` should be `_trip_schedule_selector` as defined here: https://github.com/UDST/urbanaccess/blob/2f37f8ea8337fd65d654e9506b3c76f753af22db/urbanaccess/gtfs/network.py#L180

File: urbanaccess/gtfs/network.py [hunk 3]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
+                                    input_trips_df=trips_subgdf,
+                                    input_calendar_df=gtfsfeeds_df.calendar,
+                                    day=day)
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: You are missing one required kwarg here: `input_calendar_dates_df=gtfsfeeds_dfs.calendar_dates,`

File: urbanaccess/gtfs/network.py [hunk 4]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
+
+    # confirm if need to interpolate stop times and, if yes, then do so
+    interpolate_trigger = (gtfsfeeds_df.stop_times_int.empty or 
+                           overwrite_existing_stop_times_int or 
+                           use_existing_stop_times_int == False)
+    if interpolate_trigger:
+        interp_stops = interpolatestoptimes(
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: `interpolatestoptimes` should be `_interpolate_stop_times`.

File: urbanaccess/gtfs/network.py [hunk 5]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
+                        stop_times_df=gtfsfeeds_df.stop_times,
+                        calendar_selected_trips_df=calendar_selected_trips_df,
+                        day=day)
+
+        time_diff_interp_stops = timedifference(stop_times_df=interp_stops)
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: `timedifference` should be `_time_difference`.

File: urbanaccess/gtfs/network.py [hunk 6]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
-        stop_time_df=gtfsfeeds_dfs.stop_times)
-
-    ua_network.transit_edges = _route_id_to_edge(
-        transit_edge_df=ua_network.transit_edges,
-        trips_df=gtfsfeeds_dfs.trips)
+        err_text = ('Existing stop_times_int is empty. '
+                    'set use_existing_stop_times_int to False to create it.')
+        assert gtfsfeeds_df.stop_times_int.empty == False, err_text
+
+    selected_interpolated_stop_times_df = timeselector(
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: `timeselector` should be `_time_selector`.

File: urbanaccess/gtfs/network.py [hunk 7]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
+    # generate final edges dataframe for transit
+    desired_stops_int_cols = ['unique_trip_id',
+                              'stop_id',
+                              'unique_stop_id',
+                              'timediff',
+                              'stop_sequence',
+                              'unique_agency_id',
+                              'trip_id']
+    sub_stops_int = selected_interpolated_stop_times_df[desired_stops_int_cols]
+    final_edge_table = format_transit_net_edge(stop_times_df=sub_stops_int)
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: `format_transit_net_edge` should be `_format_transit_net_edge`.

File: urbanaccess/gtfs/network.py [hunk 8]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
+    ua_network.transit_edges = convert_imp_time_units(
+                                                df=final_edge_table,
+                                                time_col='weight',
+                                                convert_to='minutes')
+
+    # similarly, finalize select transit network nodes and apply to network
+    fin_sel_stops = stops_in_edge_table_selector(
+                        input_stops_df=gtfsfeeds_df.stops,
+                        input_stop_times_df=selected_interpolated_stop_times_df)
+    ua_network.transit_nodes = format_transit_net_nodes(df=fin_sel_stops)
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: `format_transit_net_edge` should be `_format_transit_net_edge`.

File: urbanaccess/gtfs/network.py [hunk 9]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
+
+    # TODO: Undesirable to continually update networks.transit_edges
+    #       Should reach a "final" state before updating (so fix above)
+    ua_network.transit_edges = route_type_to_edge(
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: `route_type_to_edge` should be `_route_type_to_edge`.

File: urbanaccess/gtfs/network.py [hunk 10]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
+                                    transit_edge_df=ua_network.transit_edges,
+                                    stop_time_df=gtfsfeeds_df.stop_times)
+
+    ua_network.transit_edges = route_id_to_edge(
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: `route_id_to_edge` should be `_route_id_to_edge`.

File: urbanaccess/gtfs/network.py [hunk 11]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
+                              'unique_stop_id',
+                              'timediff',
+                              'stop_sequence',
+                              'unique_agency_id',
+                              'trip_id']
+    sub_stops_int = selected_interpolated_stop_times_df[desired_stops_int_cols]
+    final_edge_table = format_transit_net_edge(stop_times_df=sub_stops_int)
+
+    # now convert to desired time format and apply to global network value
+    ua_network.transit_edges = convert_imp_time_units(
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: `convert_imp_time_units` should be `_convert_imp_time_units`.

File: urbanaccess/gtfs/network.py [hunk 12]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
+                                                df=final_edge_table,
+                                                time_col='weight',
+                                                convert_to='minutes')
+
+    # similarly, finalize select transit network nodes and apply to network
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: `stops_in_edge_table_selector` should be `_stops_in_edge_table_selector`

**Comment by bhelx** [comment]
I've commented with all the changes I needed to get this to work for me. I'm on python `3.6` but I think it should work on `3.5` as well. If this doesn't get updated soon I'd be happy to share my changes as a PR.

**Comment by kuanb** [comment]
Thanks. I'd been playing with a streamlined version and had failed to account for that when updating this PR.

If you've made the changes, would you mind opening a PR against my branch?

I can accept your updates and then we can merge that into this PR.

**Comment by kuanb** [comment]
Also highlights the need for more tests! :)

**Comment by bhelx** [comment]
@kuanb sure, the PR is in your fork: https://github.com/kuanb/urbanaccess/pull/1

**Comment by sablanchard** [comment]
Thanks @bhelx and @kuanb we have not yet gotten to reviewing this PR. We wanted to insert more unit tests before merging which we have not yet had much time to address but eventually will. FYI: we are releasing a large PR by early next week to expand the number of GTFS structure types supported and this includes a bit of code refactoring.

**Comment by bhelx** [comment]
@sablanchard I've only smoke tested this and the way I'm using it doesn't cover much code. It's likely there are some other issues and possibly some that break 2.7 support. My vote would be to leave it in a forked state until the unit tests are there. Although if there is someone who heavily uses this with python3, they could bless it as safe.

**Comment by sablanchard** [comment]
Thanks @bhelx that will be the plan. Curious to hear about your use case if you wanted to share, we are always interested in how the library is being used! Expect incremental improvements in the library over this summer as time permits to get it out of alpha.

File: urbanaccess/gtfs/network.py [hunk 13]
```diff
@@ -124,55 +141,89 @@ def create_transit_net(gtfsfeeds_dfs, day,
...
-    ua_network.transit_edges = _route_type_to_edge(
-        transit_edge_df=ua_network.transit_edges,
-        stop_time_df=gtfsfeeds_dfs.stop_times)
-
-    ua_network.transit_edges = _route_id_to_edge(
-        transit_edge_df=ua_network.transit_edges,
-        trips_df=gtfsfeeds_dfs.trips)
+        err_text = ('Existing stop_times_int is empty. '
+                    'set use_existing_stop_times_int to False to create it.')
+        assert gtfsfeeds_df.stop_times_int.empty == False, err_text
```
Comment (bhelx) [inline urbanaccess/gtfs/network.py]: I think i missed this one in my PR

**Pull request closed without merging**